File: gcc/testsuite/libdiagnostics.dg/test-labelled-ranges.c
/* Example of multiple locations, with labelling of ranges.

   Intended output is similar to:

PATH/test-labelled-ranges.c:9:6: error: mismatching types: 'int' and 'const char *'
   19 |   42 + "foo"
      |   ~~ ^ ~~~~~
      |   |    |
      |   int  const char *

   along with the equivalent in SARIF.  */

#include "libdiagnostics.h"
#include "test-helpers.h"

/*
_________11111111112
12345678901234567890
  42 + "foo"
*/
const int line_num = __LINE__ - 2;

int
main ()
{
  begin_test ("test-labelled-ranges.c.exe",
	      "test-labelled-ranges.c.sarif",
	      __FILE__, "c");

  const diagnostic_physical_location *loc_operator
    = diagnostic_manager_new_location_from_file_line_column (diag_mgr,
							     main_file,
							     line_num,
							     6);

  /* begin quoted source */
  diagnostic *d = diagnostic_begin (diag_mgr,
				    DIAGNOSTIC_LEVEL_ERROR);
  diagnostic_set_location (d, loc_operator);
  diagnostic_add_location_with_label (d,
				      make_range (diag_mgr,
						  main_file,
						  line_num, 3, 4),
				      "int");
  diagnostic_add_location_with_label (d,
				      make_range (diag_mgr,
						  main_file,
						  line_num, 8, 12),
				      "const char *");
  
  diagnostic_finish (d, "mismatching types: %qs and %qs", "int", "const char *");
  /* end quoted source */
  
  return end_test ();
}

/* Check the output from the text sink.  */
/* { dg-regexp "\[^\n\r\]+test-labelled-ranges.c:19:6: error: mismatching types: 'int' and 'const char \\*'" } */
/* { dg-begin-multiline-output "" }
   19 |   42 + "foo"
      |   ~~ ^ ~~~~~
      |   |    |
      |   int  const char *
   { dg-end-multiline-output "" } */

/* Verify that some JSON was written to a file with the expected name:
   { dg-final { verify-sarif-file } } */

/* Use a Python script to verify various properties about the generated
   .sarif file:
   { dg-final { run-sarif-pytest test-labelled-ranges.c "test-labelled-ranges.py" } } */

File: gcc/testsuite/libdiagnostics.dg/test-labelled-ranges.py
# Verify the SARIF output of test-labelled-ranges.{c,cc}

from sarif import *

import pytest

@pytest.fixture(scope='function', autouse=True)
def sarif():
    return sarif_from_env()

def test_sarif_output(sarif):
    schema = sarif['$schema']
    assert schema == 'https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/schemas/sarif-schema-2.1.0.json'

    version = sarif['version']
    assert version == '2.1.0'

    runs = sarif['runs']
    run = runs[0]

    results = run['results']
    assert len(results) == 1
    assert results[0]['ruleId'] == 'error'
    assert results[0]['level'] == 'error'
    assert results[0]['message']['text'] \
        == "mismatching types: 'int' and 'const char *'"
    assert len(results[0]['locations']) == 1
    location = results[0]['locations'][0]
    phys_loc = location['physicalLocation']
    assert phys_loc['region']['startLine'] == 19
    assert phys_loc['region']['startColumn'] == 6
    assert phys_loc['region']['endColumn'] == 7
    assert phys_loc['contextRegion']['startLine'] == 19
    assert phys_loc['contextRegion']['snippet']['text'] \
        == '  42 + "foo"\n'

    annotations = location['annotations']
    assert len(annotations) == 2

    assert annotations[0]['startLine'] == 19
    assert annotations[0]['startColumn'] == 3
    assert annotations[0]['endColumn'] == 5
    assert annotations[0]['message']['text'] == 'int'

    assert annotations[1]['startLine'] == 19
    assert annotations[1]['startColumn'] == 8
    assert annotations[1]['endColumn'] == 13
    assert annotations[1]['message']['text'] == 'const char *'
